### PR TITLE
Test suite for Python SDK (Dont merge this)

### PR DESCRIPTION
**Pull request by nghiant2710**
This is how we organize the test suite for Python SDK (there are only test cases for `application` class atm but more will be added soon).

This will fix a part of https://github.com/resin-io/resin-sdk-python/issues/39.

### CODE REVIEW

**Comment by lurch** [comment]
@pimterry Does dropping support for Python 2.6 count as a "breaking" (and hence major) change?

File: resin/base_request.py
Comment (lurch) [inline resin/base_request.py]: Should this be named something like `decode_utf8` rather than `to_utf` ?

File: resin/base_request.py
Comment (lurch) [inline resin/base_request.py]: Just out of curiosity, where are these `'bytes'` values coming from?

File: tests/functional/models/test_application.py
Comment (lurch) [inline tests/functional/models/test_application.py:76]: Is having to do `json.loads(some_api_call().decode('utf-8'))` a "code smell" that `some_api_call()` isn't operating at the right level of abstraction?  (i.e. would it make more sense for the `decode('utf-8')` and `json.loads()` to be done _inside_ `some_api_call()`? Although this would obviously be a breaking change)

File: tests/functional/models/test_application.py
Comment (lurch) [inline tests/functional/models/test_application.py]: Curiously, the comment a few lines above claims that the application name has to be >= 3 characters, and yet the error message that you're checking implies that the application name has has to be >= 4 characters? They can't both be correct ;)

File: tests/functional/models/test_application.py
Comment (lurch) [inline tests/functional/models/test_application.py:81]: This `app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))` is being done so many times, that perhaps it's worth moving it to e.g. a `create_test_application(self)` function?

EDIT: e.g.
```python
def create_test_application(self, name='FooBar', device_type='Raspberry Pi 2'):
    return json.loads(self.resin.models.application.create(name, device_type).decode('utf-8'))
```
as that'd allow you to just use `app = self.create_test_application()` in most places, while still allowing the `name` and `device_type` to be specified as needed in other tests.

File: tests/functional/models/test_application.py
Comment (lurch) [inline tests/functional/models/test_application.py:104]: With all these enable and disable tests, I wonder if it _might_ make more sense to do e.g.
```python
    def test_rolling_updates(self): # or test_toggle_rolling_updates
        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
        # I guess at this point you could also test the default value of app['should_track_latest_release'] ?
        # should enable rolling update for the applications devices.
        self.resin.models.application.enable_rolling_updates(app['id'])
        app = self.resin.models.application.get('FooBar')
        self.assertTrue(app['should_track_latest_release'])
        # should disable rolling update for the applications devices.
        self.resin.models.application.disable_rolling_updates(app['id'])
        app = self.resin.models.application.get('FooBar')
        self.assertFalse(app['should_track_latest_release'])
```
?
And if the default value of `app['should_track_latest_release']` is True, then it'd probably make sense to test disabling it before testing enabling it.

File: tests/functional/models/test_application.py
Comment (lurch) [inline tests/functional/models/test_application.py]: Might make sense to tweak this to:
```python
        support_expiry_time = int((support_date - datetime.utcfromtimestamp(0)).total_seconds() * 1000)
        self.assertEqual(support_expiry_time, expiry_time)
```
(IMHO it makes it slightly more obvious what's going on, and this'll probably be easier to debug should any problems arise)

File: tests/helper.py
Comment (lurch) [inline tests/helper.py]: @pimterry Do we really want to be storing resin account credentials in plaintext? How are other API / SDK tests done?
(I guess anybody could log in with the details above, change the account's password, and then suddenly all the PythonSDK CI tests would break)

File: tests/helper.py
Comment (lurch) [inline tests/helper.py:75]: This isn't used anywhere (yet) ?

File: tests/functional/models/test_application.py
Comment (lurch) [inline tests/functional/models/test_application.py]: As there's only one application, you could also check the `app_name`, as you do in the next test?

Would it also be worth creating a test for adding two applications, to check that `get_all()` can cope with more than one result? ;) And you could check e.g.
```python
all_applications = self.resin.models.application.get_all()
self.assertEqual(len(all_applications), 2)
self.assertNotEqual(all_applications[0]['app_name'], all_applications[1]['app_name'])
```

File: tests/functional/models/test_application.py
Comment (lurch) [inline tests/functional/models/test_application.py]: Is this just `Exception` and not some more specific type?

File: tests/functional/models/test_application.py
Comment (lurch) [reply tests/functional/models/test_application.py]: Or in fact, perhaps adding
```python
def datetime_to_epoch_ms(datevalue):
    return int((datevalue - datetime.utcfromtimestamp(0)).total_seconds() * 1000)
```
would be useful, as there are several places in these unit-tests where that could be used :-)

File: tests/functional/models/test_application.py
Comment (lurch) [reply tests/functional/models/test_application.py:76]: In fact, this is actually an inconsistency in the docs as https://github.com/resin-io/resin-sdk-python/blob/master/DOCUMENTATION.md#function-createname-device_type says "Returns: dict: application info." but then the example shows it just returning a single string. Should I create a separate issue for this?

**Review by lurch — COMMENTED** [review]
Left 'a few' comments, but this looks like a great start to the test suite :+1:

File: tests/functional/models/test_application.py
Comment (lurch) [reply tests/functional/models/test_application.py]: Haha, which is also what https://stackoverflow.com/a/11111177 says ;)

File: .travis.yml
Comment (pimterry) [inline .travis.yml]: Nit: Stye -> Style

File: tests/functional/models/test_application.py
Comment (pimterry) [reply tests/functional/models/test_application.py:76]: This is definitely a bit awkward, and good to fix, though it's fine to do in a later PR

File: tests/functional/models/test_application.py
Comment (pimterry) [reply tests/functional/models/test_application.py:104]: Personally I'd rather have more & smaller tests - makes them easier to read, easier to tell what exactly is broken if they fail, and easier to run separate bits in isolation.

File: tests/helper.py
Comment (pimterry) [reply tests/helper.py]: Indeed not. The SDK credentials read from the environment, loading a `.env` file automatically if present, so you just create a `.env` with credentials to run tests locally, and we set the env vars in the Travis/etc build config. We should do that here too (and reset these login details), both for security, and because that way you can use different login details locally/in CI. If you use the same login details everywhere then running the tests locally will break any running CI builds.

File: tests/helper.py
Comment (pimterry) [inline tests/helper.py:34]: :+1: :+1: :+1: Good move! I need to do this for the JS tests too.

File: tests/helper.py
Comment (pimterry) [inline tests/helper.py]: This looks like a debugging leftover, should probably be removed

File: tests/helper.py
Comment (pimterry) [reply tests/helper.py:75]: Yep. It should be - you probably just want a single `reset_all` method that includes both, and to then run that from the tests instead.

**Review by pimterry — CHANGES_REQUESTED** [review]
A few comments, but this looks like a great start. The one main thing we need to resolve I think is getting credentials from the env instead of hardcoding them.

The point about breaking changes is an interesting one - if this PR is actually removing support for Python 2.6 somehow, this is a breaking change. If we're stopping advertising it but it will keep working at least until the next major bump, or if it's actually never worked at all for the current major version and it's a documentation bug, then it isn't.

File: tests/functional/models/test_application.py
Comment (lurch) [reply tests/functional/models/test_application.py:104]: Fair enough. Should there also be tests for the default values? And presumably if the default value is e.g. True, then when testing the "enable" function you'd first need to disable it, to ensure that the enable isn't just a NOP?

File: tests/functional/models/test_application.py
Comment (nghiant2710) [reply tests/functional/models/test_application.py:76]: Created an issue for this https://github.com/resin-io/resin-sdk-python/issues/68. I couldn't remember why I made it like that but we need to fix this.

File: tests/functional/models/test_application.py
Comment (nghiant2710) [reply tests/functional/models/test_application.py:81]: As mentioned above, I will fix the return value for this function so will update the test cases in the PR for this issue https://github.com/resin-io/resin-sdk-python/issues/68 later.

File: tests/functional/models/test_application.py
Comment (nghiant2710) [reply tests/functional/models/test_application.py:104]: @lurch, I don't think we need a test for default values here but it makes sense to enable/disable first before testing disable/enable function to ensure our test is doing a right thing.

File: tests/functional/models/test_application.py
Comment (nghiant2710) [reply tests/functional/models/test_application.py]: 👍 Updated

File: tests/helper.py
Comment (nghiant2710) [reply tests/helper.py]: Fixed. I've created another PR just to see the test result https://github.com/resin-io/resin-sdk-python/pull/69 since travis doesn't allow PR from downstream repo to access encrypted variables.

File: tests/helper.py
Comment (nghiant2710) [reply tests/helper.py:75]: @lurch @pimterry since these teardown functions are run after every test case so I want to keep it as small as possible, really don't want to spam the API with unnecessary requests. The fully reset function isn't used yet but sure we will use it for coming test case.

File: tests/helper.py
Comment (nghiant2710) [reply tests/helper.py]: Updated

**Comment by nghiant2710** [comment]
@lurch @pimterry, I know dropping support for Python 2.6 is a breaking change but on the other hand, I'm pretty sure that no one is using the SDK with Python 2.6 so this change won't break anyone application. Moreover, we want the test suite landed as soon as possible (Don't want to wait until next major version is merged) so I make it a `minor` change.

Anyway, I don't have a strong opinion about this so if we agree this is a breaking change then I will try to finish the while test suite soon and then bump major version.

File: resin/base_request.py
Comment (nghiant2710) [reply resin/base_request.py]: updated

File: resin/base_request.py
Comment (nghiant2710) [reply resin/base_request.py]: In Python 3, a bytes string can't be dumped into JSON object so we need to encode any bytes strings using the encoding `utf-8` first.

File: tests/functional/models/test_application.py
Comment (nghiant2710) [reply tests/functional/models/test_application.py]: Yeah, we want to validate the error message here so a general exception is fine with me.

File: tests/functional/models/test_application.py
Comment (nghiant2710) [reply tests/functional/models/test_application.py]: Fixed. I copied from the JS test so we probably need to fix it here as well https://github.com/resin-io/resin-sdk/blob/master/tests/integration/models/application.spec.coffee#L59-L64

**Comment by pimterry** [comment]
@nghiant2710 if people are using this in 2.6, would this PR break their code?

**Comment by nghiant2710** [comment]
@pimterry, it won't break their code, just the test suite won't be able to run on Python 2.6

File: tests/functional/models/test_application.py
Comment (nghiant2710) [reply tests/functional/models/test_application.py]: added this to the test case.

**Comment by pimterry** [comment]
@nghiant2710 I think that's totally fine then, and this is just a minor change.

As long as their code still works, everything's ok, and all we're really doing with this PR is changing which version support we explicitly advertise.

File: tests/helper.py
Comment (pimterry) [inline tests/helper.py]: This will never be thrown, as `os.environ.get` doesn't throw if the variable doesn't exist (it just returns `None`)

**Review by pimterry — CHANGES_REQUESTED** [review]
One last small bug in the credential loading. The rest all looks good to me.

File: tests/helper.py
Comment (nghiant2710) [reply tests/helper.py]: @pimterry updated, by doing `os.environ['Key']`, it will raise `KeyError` exception, if the environment variable is not present.

**Review by pimterry — APPROVED** [review]
All the code looks good to me now. The build is failing though because env vars aren't provided to PRs from external sources, only PRs on the repo itself. It looks like you've submitted this from a PR on your own branch?

That's unusual - I think the standard Resin process is to open branches directly on resin repo itself, and PR from there. You'll need to move to that approach to make this work. Otherwise from Github's point of view you're an unknown external contributor, so you shouldn't be given free access to the credentials.

That'll mean reopening this PR, but it's otherwise easy: push this same branch directly to this repo instead, open a new PR with that, and I'll approve so you can merge.

**Comment by nghiant2710** [comment]
@pimterry, I noticed that travis won't include encrypted env vars if the PR is from external sources, that's why i created this PR to show the test result https://github.com/resin-io/resin-sdk-python/pull/69.

From now on, I will create PR from branches belong to resin repo.

**Comment by nghiant2710** [comment]
@hedss, I already pushed the v2.1.1 release tags, do you know why VersionBot still reports this error?

**Comment by lurch** [comment]
I think for VB to be able to do the merging, the PR needs to be coming from the same repo and not from the "external" `nghiant2710/resin-sdk-python` repo? Might be easiest to close this PR, and reapply the commits on a branch from `resin-io/resin-sdk-python` ?

**Pull request closed without merging**
